### PR TITLE
Domain transfer lock API

### PR DESCRIPTION
**Pull request by DXTimer**
This PR adds OpenAPI definitions for the new Domain Transfer Lock API and its related events.

There are three endpoints that are added:

1. `GET /:account/registrar/domains/:domain/transfer_lock`
2. `POST /:account/registrar/domains/:domain/transfer_lock`
3. `DELETE /:account/registrar/domains/:domain/transfer_lock`

Two already existing but undocumented events have been added:

1. `domain.lock`
2. `domain.unlock`

Belongs to https://github.com/dnsimple/dnsimple-business/issues/1728

### CODE REVIEW

File: content/v2/openapi.yml [hunk 1]
```diff
@@ -1287,6 +1287,75 @@ paths:
...
       summary: De-delegate from vanity name servers
+  '/{account}/registrar/domains/{domain}/transfer_lock':
+    get:
+      description: Gets the transfer lock status for a domain.
+      parameters:
+        - $ref: '#/components/parameters/Account'
+        - $ref: '#/components/parameters/Domain'
+      operationId: getDomainTransferLock
+      tags:
+        - registrar
```
Comment (wilson-dnsimple) [inline content/v2/openapi.yml:1298]: We should delete this tag as per https://github.com/dnsimple/dnsimple-developer/pull/523

File: content/v2/openapi.yml [hunk 2]
```diff
@@ -1287,6 +1287,75 @@ paths:
...
+        '404':
+          $ref: '#/components/responses/404'
+    delete:
+      description: Unlocks the domain to allow domain transfers.
+      parameters:
+        - $ref: '#/components/parameters/Account'
+        - $ref: '#/components/parameters/Domain'
+      operationId: disableDomainTransferLock
+      tags:
+        - registrar
```
Comment (wilson-dnsimple) [inline content/v2/openapi.yml:1343]: Same as above

File: content/v2/openapi.yml [hunk 3]
```diff
@@ -1287,6 +1287,75 @@ paths:
...
+          $ref: '#/components/responses/404'
+      summary: Domain transfer lock status
+    post:
+      description: Locks the domain to prevent unauthorized transfers.
+      parameters:
+        - $ref: '#/components/parameters/Account'
+        - $ref: '#/components/parameters/Domain'
+      operationId: enableDomainTransferLock
+      tags:
+        - registrar
```
Comment (wilson-dnsimple) [inline content/v2/openapi.yml:1320]: Same as above

File: content/v2/openapi.yml [hunk 4]
```diff
@@ -3931,6 +4007,12 @@ components:
       properties:
         domain:
           $ref: '#/components/schemas/Domain'
+    EventDomainLock:
```
Comment (wilson-dnsimple) [inline content/v2/openapi.yml:4010]: This doesn't seem to be used?

File: content/v2/openapi.yml [hunk 5]
```diff
@@ -3949,6 +4031,12 @@ components:
       properties:
         domain:
           $ref: '#/components/schemas/Domain'
+    EventDomainUnlock:
```
Comment (wilson-dnsimple) [inline content/v2/openapi.yml:4034]: This doesn't seem to be used?

**Review by wilson-dnsimple — APPROVED** [review]

File: content/v2/openapi.yml [hunk 6]
```diff
@@ -3931,6 +4006,12 @@ components:
       properties:
         domain:
           $ref: '#/components/schemas/Domain'
+    EventDomainLock:
+      type: object
+      description: Payload for domain.lock
```
Comment (weppos) [inline content/v2/openapi.yml:4011]: `domain.lock` is too generic. Note we had the same issue in the actual interactor code, and [they were renamed](https://github.com/dnsimple/dnsimple-app/pull/23326).

Please use something that is more closely associated with the notion of transfer lock, for instance:

- `domain.transfer_lock_enable`
- `domain.transfer_lock_disable`

File: content/v2/webhooks/events.markdown [hunk 7]
```diff
@@ -90,6 +90,7 @@ The following events are available:
 * domain.auto\_renewal\_enable
 * domain.create
 * domain.delete
+* domain.lock
```
Comment (weppos) [inline content/v2/webhooks/events.markdown:93]: Another reason for the rename. Notice how events are significantly apart which makes it even less readable.

File: content/v2/webhooks/events.markdown [hunk 8]
```diff
@@ -90,6 +90,7 @@ The following events are available:
...
 * domain.register:started
```
Comment (weppos) [inline content/v2/webhooks/events.markdown:94]: Not related to this PR, but when changes were made in https://github.com/dnsimple/dnsimple-developer/pull/505 it seems we did not keep names sorted properly. Can you please have state-aware events (the ones ending with `:`) added at the end?

```
* domain.transfer
* domain.transfer:started
* domain.transfer:cancelled
```

vs

```
* domain.transfer:started
* domain.transfer
* domain.transfer:cancelled
```

This for all cases in this file.

File: fixtures/v2/webhooks/domain.lock/example.http [hunk 9]
```diff
@@ -0,0 +1,9 @@
+POST / HTTP/1.1
+Host: example.com
+Accept-Encoding: gzip
+Content-Type: application/json
+User-Agent: DNSimple-Webhook-Notifier/8fd98a84516b72d5b863b27f93fcbaef61f06c03
+Content-Length: 643
+Connection: keep-alive
+
+{"data": {"domain": {"id": 1, "name": "example.com", "state": "registered", "account_id": 1010, "auto_renew": false, "created_at": "2023-03-02T02:39:18Z", "expires_at": "2024-03-02T02:39:22Z", "expires_on": "2024-03-02", "updated_at": "2023-08-31T06:46:48Z", "unicode_name": "example.com", "private_whois": false, "registrant_id": 101}}, "name": "domain.lock", "actor": {"id": "1010", "entity": "account", "pretty": "xxxxxxx-xxxxxxx-xxxxxxx@xxxxx.com"}, "account": {"id": 1010, "display": "xxxxxxx-xxxxxxx-xxxxxxx", "identifier": "xxxxxxx-xxxxxxx-xxxxxxx@xxxxx.com"}, "api_version": "v2", "request_identifier": "0f31483c-c303-497b-8a88-2edb48aa111e"}
```
Comment (weppos) [inline fixtures/v2/webhooks/domain.lock/example.http:9]: **Note for the future:**

It's a bit awkward that the payload doesn't have any information regarding the lock status. Generally, the payload carries over details about the change.

As we currently can't include the lock status in the domain (for the performance reason we discussed), we would need to add a separate, specific node just for it. Given the payload name carries over the status, I'd leave it as it is.

Should we ever be able to sync the lock status locally in the domain object, we should definitely expose it here.

/cc @OleMchls for knowledge about the current limitations imposed by not storing the status locally

**Review by weppos — CHANGES_REQUESTED** [review]
Please have a look at the notes inline.

File: content/v2/openapi.yml (hunk 6 above)
Comment (AGS4NO) [reply content/v2/openapi.yml]: @weppos I have renamed the event and payload items per your suggestion.

File: content/v2/webhooks/events.markdown (hunk 8 above)
Comment (AGS4NO) [reply content/v2/webhooks/events.markdown]: These have all been alphabetized with state-aware events at the end of each section.

```
* domain.transfer
* domain.transfer:cancelled
* domain.transfer:started
```

File: content/v2/openapi.yml (hunk 6 above)
Comment (weppos) [reply content/v2/openapi.yml]: @AGS4NO did you use "enable" or "enable**d**" (notice the -ed form). For consistency with all the other events, we should use "enable/disable"

Comment (weppos) [reply content/v2/openapi.yml]: Never mind, I was looking an outdated diff where it used the -ed form, but it sounds like it is now fixed.

**Review by weppos — APPROVED** [review]

**Pull request merged**